### ISSUE TITLE
Feature: put sequences into canonical format

### ISSUE THREAD

**Issue by mpogue2** [mpogue2/SquareDesk#488]
This is a prerequisite of colored card work (#487) and more generally sequence design work.  I am putting all of the sequences (starting with SS -> Resolve) from ceder dot net into a canonical form, so that they can be analyzed, imported, etc..

**Comment by danlyke** [comment]
When you've got an example (of 1 is fine), I'd love to have it thrown at me. Of course it may be months before I get back here, but I'm looking at this today...

**Comment by mpogue2** [comment]
See also #101 .  What I'm finding is that the Ceder data is exceedingly "dirty", as in "inconsistent", or "wrong", etc....  I am cleaning that up right now...

Here are some of the things I've corrected so far (I'm doing the correction outside of SquareDesk, because it's far easier):

- Multiple spellings for the same word, e.g. "Right & Left Thru" vs "Right And Left Thru"
- Multiple calls on one line (I am splitting them into multiple lines), e.g. "Pass The Ocean & Swing Thru"
- Extra words that are not needed, e.g. "Boys Only Zoom"
- Inconsistent use of "&" vs "And".  I am cleaning this up, e.g. "Right And Left Thru", but "Follow Your Neighbor & Spread".

A lot of this is similar to the singing call analysis code I wrote last year, where I put everything into a single line in Rich Reel's format, and then did statistical analyses on the resulting lines.

I am still cleaning, but I'll post an R file when I'm done.  My intent is to clean up all of the SS->Resolves in Vic's library of sequences (basic, MS, Plus, A1, A2 only) into a single canonical format.  I can also post the results of that cleanup when done!  You could then think about importing that format into a sequence editor (I'm also quite interested in making colored cards, which are a BIG hit down here...).

Right now, I have about 783 sequences in the process of being cleaned.  When done, this should work for cleaning non-SS->Resolve sequences, too.

**Comment by mpogue2** [comment]
I am also splitting off the "who" and the "Suffix", e.g.
Boys Follow Your Neighbor & Spread ->
who: "Boys"
call: "Follow Your Neighbor"
suffix: "& Spread"
in case that is useful to you.  I'll provide that as CSV-with-a-header, with each of those being 1 column.

**Comment by danlyke** [comment]
Cool. Part of where I'm going is that while all the metadata is awesome, and if we can get it we should have it, just handling text chunks is gonna be necessary. I'd really like to be able to not have to carry my big binder of A2 next time I call for Diablo Dancers, just use my laptop...

**Comment by mpogue2** [comment]
I have the metadata available, too.  Author, ID#, Level, Difficulty (although Difficulty is not set on many of the higher level sequences).

**Comment by mpogue2** [comment]
Note: Canonical format will also have extra spaces, extra characters, and parenthetical comments removed.  (I'm considering having a way to put the parentheticals back in, but that's a stretch goal).

In the short term, I am hoping to write a quick and dirty "give me all the class level sequences", where class level is something like "GDP minus Eight Chain and Pass to the Center", which is what I needed for the last dance I called.  It's super-painful to have to examine every sequence for class level.

That will then feed into "turn this list into 4x6 colored cards" a la #487.

Also, I am hoping that this will be suitable for automatic validation in SD!  No more hand-validation (OK, I do it by voice, but it's still painful).

**Comment by mpogue2** [comment]
OK, sent you some files to play with.  It's a LOT cleaner than it was, but still quite a bit of work to do, especially at the higher levels (which tend to do wackier things).  I'm intrigued about the possibility of calling from the laptop, and I'm looking forward to seeing what you come up with!  I'm using Notes on iOS right now, and just manually scrolling.

**Comment by danlyke** [comment]
Thanks. And I had an "aha" yesterday: I was thinking about putting all
this stuff in the database, but realized that I actually think I want
it in text files, on the filesystem.

So a start will be just being able to effectively search this stuff
in-place.

**Comment by mpogue2** [comment]
I'm not sure that putting stuff on-disk makes it very searchable.  Ideally, I think we want to do queries like:  "Find all the sequences that have a Spin the Top in them, that are Class level (MS + 8 Plus calls)".  I'm not sure how this can be done efficiently, other than using a database to make the search faster....

One option might be to slurp in all the on-disk stuff into a database, to make searching quicker.  But then when I edit something on-disk there's the "on-disk is not in sync with the database" problem.  This latter problem might be better to deal with than the "all searches take too long" problem that everybody might hit with the first approach.

**Comment by danlyke** [comment]
Yeah, these days with SSDs it seems like a simple grep-like linear search may be fast enough.

**Comment by mpogue2** [comment]
I am also thinking that it would be nice to put canonical format versions of the ceder.net sequences into (say) the sequences directory.  As long as they are searchable, I'm good.  Note that if we stick them on disk in text files, they might not be able to do be searchable using the complex queries I am thinking of (like the example above).  If so, we could have the files on disk, slurp them in at startup time (if they exist), and search using an internal database.  They could be checked for canonicalness at that time (with warnings if not?)  Also, in this case, if anybody sticks a file in that directory after the sequences have been slurped in, we should probably re-slurp, with a FileWatcher.

Probably some more specific search use cases would be good to do.  I'll come up with a few when I get a chance, but that one above is probably a typical one that I do (that is painful to do by hand).  I'm interested in what kinds of searches you are thinking of, too!

**Comment by danlyke** [comment]
I think, unfortunately, that I blew away my prototype on this thinking
it'd be easy to re-create, and here I am not yet recreating it, but...

the big search thing I was thinking of was what I had in the prototype:
Show me every sequence with the call I'm looking for, that does not
contain the calls that aren't checked in the taught calls list (with
some aggressive aliases and substrings for those calls).

This generally gave me a way to filter for singing call figures that
only contained taught calls.

**Comment by mpogue2** [comment]
I'm looking for the same thing, but more generally for sequences (not just singing calls):  show me those that have THIS call, that don't use any calls that are not on THIS list.  I would like to see us translate to a canonical format, which I've been working on (won't be any surprises, I think), so that the ingestion is fuzzy, but the search does not have to be.

Here's the ingestion file, that operates on downloaded text files to make them canonical (and then to make colored index cards of a specific size).  Most of the regex stuff can be moved to C++ in a straightforward way, I think, at ingestion time.

[process_V1.5.R.zip](https://github.com/mpogue2/SquareDesk/files/3342507/process_V1.5.R.zip)
